### ISSUE TITLE
Only include a static resource once per request, even if referenced by multiple fragments

### ISSUE THREAD

**Issue by dazld** [zalando/tailor#163]
Hopefully title is self-explanatory, but given the following fragments:

FragA -> style1.css, style2.css, script1.js, script2.js
FragB -> style2.css, style3.css, script2.js, script3.js

then I would expect to see only the following files loaded by the browser:
style1.css, style2.css, style3.css
script1.js, script2.js, script3.js

**Comment by vigneshshanmugam** [comment]
Browser also does the job of avoiding the downloads(script/style) if the file is already loaded in the current navigation. 

If script2.js from fragment 1 and script2.js from fragment 2 has the same URL, then the script would only be downloaded once.

**Comment by dazld** [comment]
Correct - but each time the script is referenced, it's re-executed in its entirety. ie: https://codepen.io/dazld/pen/zzBKwV - same script inserted a bunch of times in head. I'm trying really hard to think why someone would want to do that with tailor, but coming up a blank. It feels like scripting from a different time.

I'm not sure what CSS does, but it wouldn't surprise me if it's similar.

**Comment by addityasingh** [comment]
@dazld But IMO this is a valid requirement as we use the same fragment multiple times on a page for recommendations on PDP page and therein the same resources are references multiple times

**Comment by vigneshshanmugam** [comment]
@dazld Correct, the script would only be downloaded once and re executed at all places where it is referenced. 

Thinking out load, Not referencing the script to browser will cause these issues

1.  If fragment-1 is referenced twice in the template and wants to maintain a global state in script1.js and update the state on each exec call. This can also be an event bus trigger. 
```
// script1.js - 1st exec
var ele = doc.getElementById("fragment-a");
if (ele) {
eventBus.trigger("fragment-a-loaded");
}

// script1.js - 2nd exec
var ele = doc.getElementById("fragment-b");
if (ele) {
eventBus.trigger("fragment-b-loaded");
}
```
This is totally valid and people might even have tracking based on this. 

2. render will never be called for the second time which means if fragment-1 is referenced twice in the template. Only the first one will render, all others will be ignored because the script that has the render function will never be executed.

**Comment by iilei** [comment]
IMO that should stay a concern of the fragments which trigger loading identical scripts.

**Comment by dazld** [comment]
that sounds like a buggy implementation - relying on the browser to execute the same script instead of cleanly separating the behaviour to a single place and invoking it multiple times.

**Comment by addityasingh** [comment]
@dazld Do you mean that we dedupe the fragment scripts if the fragment is repeated on the page, and just call it? If yes, Should that be the responsibility of Tailor? IMO it should be the responsibility of fragment owners to do that

**Comment by vigneshshanmugam** [comment]
@dazld You cant say its a bugg one. If we have recommendations fragment on the top and bottom of the page. Its valid to send the same script twice.

**Comment by dazld** [comment]
dependency handling is best left to dependency handling systems. I don't think injecting the same script tag twice is a great substitute, but I guess it works in a pinch.

**Comment by iilei** [comment]
>  I don't think injecting the same script tag twice is a great substitute

referencing it twice (by two fragments referencing it) should result in requesting it twice (taking browser caching out of the equation for the sake of brevity), anything else would violate SoC and would be invasive.

I think @dazld your use case might be served best with a Singleton pattern for the script which is prone to be referenced multiple times. It would come with the benefit of being expressive.

**Comment by dazld** [comment]
true, the architecture I have in mind is opinionated. it's definitely interesting to explore, but I get the impression that I've not expressed it well as yet, my apologies for this. 

here's a plan - I'll implement a scratch approach with an example, and then there's something concrete to pick apart.

**Comment by vigneshshanmugam** [comment]
@dazld Did you have time to check to explore on the prototype implementation?

**Comment by dazld** [comment]
sorry, been totally snowed under. I still think the base idea is sound, though.

**Comment by JaroVDH** [comment]
I tried out the examples, and I noticed that scripts only get included once, but stylesheets are still added per fragment.
Why was this distinction made?

**Comment by vigneshshanmugam** [comment]
@JaroVDH There is no distinction as such, Can you give an example?

**Comment by vigneshshanmugam** [comment]
You would see the same for javascript, If you do view source of that file. In the end, You wont see two downloads, thats what matters. 
<img width="915" alt="screen shot 2018-10-19 at 12 36 07" src="https://user-images.githubusercontent.com/3902525/47213616-d9119580-d39b-11e8-90eb-8527c7899ae9.png">
